src/base/base.c

```diff
@@ -874,18 +884,36 @@ static bool should_run_user_handler_wrapper(zend_execute_data *execute_data)
 #endif
 }
 
+void xdebug_save_statement_handler(zend_extension *extension, statement_handler_func_t statement_handler)
+{
+	xdebug_extension = extension;
+	existing_statement_handler = statement_handler;
+}
+
+
+static void xdebug_disable_statement_extension_handler()
+{
+	existing_statement_handler = xdebug_extension->statement_handler;
+	xdebug_extension->statement_handler = NULL;
+}
+
+static void xdebug_enable_statement_extension_handler()
+{
+	xdebug_extension->statement_handler = existing_statement_handler;
+}
+
 /* We still need this to do "include", "require", and "eval" */
 static void xdebug_execute_ex(zend_execute_data *execute_data)
 {
 	bool run_user_handler = should_run_user_handler_wrapper(execute_data);
 
-	if (run_user_handler) {
+	if (run_user_handler && (XG_DBG(debugger_disabled) == 0 || execute_data->func->type == ZEND_EVAL_CODE)) {
```

Comment:
We only call the xdebug_execute_user_code_begin() and xdebug_execute_user_code_end() functions if the debugger has not been disabled. Notice that we cannot just use a single if statement for both functions because the status of the debugger_disabled flag can change during the call to the xdebug_old_execute_ex() function

We also need to call the begin function if we are entering into some eval code because we want to record this eval code so that it can be used later